### PR TITLE
[SPARK-36376][SQL] Collapse repartitions if there is a project between them

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Collapse repartitions if there is a project between them. For example:
```scala
sql("SELECT a FROM (SELECT id AS a, id AS b FROM range(10) DISTRIBUTE BY a, b) t DISTRIBUTE BY a").explain(true)
```
Before this pr:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Exchange hashpartitioning(a#0L, 5), REPARTITION_BY_COL, [id=#18]
   +- Project [a#0L]
      +- Exchange hashpartitioning(a#0L, b#1L, 5), REPARTITION_BY_COL, [id=#16]
         +- Project [id#2L AS a#0L, id#2L AS b#1L]
            +- Range (0, 10, step=1, splits=2)
```

After this pr:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Exchange hashpartitioning(a#0L, 5), REPARTITION_BY_COL, [id=#10]
   +- Project [id#2L AS a#0L]
      +- Range (0, 10, step=1, splits=2)
```

### Why are the changes needed?

Reduce shuffle to improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by cloud-fan** [comment]
can we come up with a general solution? or do we have a list of operators that can have `Project` between them but we can still collapse?

**Comment by wangyum** [comment]
It seems this is the last rule needs to be considered the `Project`.

These rules may include `Project`: `RemoveRedundantAggregates`, `CombineFilters` and `PushPredicateThroughJoin`, but they will all be optimized by other rules. For example:
```sql
create table t1(a int, b int, c int) using parquet;
create table t2(x int, y int, z int) using parquet;
```
```
select x, y from (select a as x, b as y, c as z from t1 group by a, b, c) t group by x, y

=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CollapseProject ===
 Aggregate [x#5, y#6], [x#5, y#6]                            Aggregate [x#5, y#6], [x#5, y#6]
!+- Project [x#5, y#6]                                       +- Aggregate [a#0, b#1, c#2], [a#0 AS x#5, b#1 AS y#6]
!   +- Aggregate [a#0, b#1, c#2], [a#0 AS x#5, b#1 AS y#6]      +- Relation default.t1[a#0,b#1,c#2] parquet
!      +- Relation default.t1[a#0,b#1,c#2] parquet 
```

```
select x, y from (select a as x, b as y, c as z from t1 where a > 1) t where y > 1

=== Applying Rule org.apache.spark.sql.catalyst.optimizer.PushDownPredicates ===
 Project [x#0, y#1]                                     Project [x#0, y#1]
!+- Filter (y#1 > 1)                                    +- Project [a#3 AS x#0, b#4 AS y#1, c#5 AS z#2]
!   +- Project [a#3 AS x#0, b#4 AS y#1, c#5 AS z#2]        +- Filter ((a#3 > 1) AND (b#4 > 1))
!      +- Filter (a#3 > 1)                                    +- Relation default.t1[a#3,b#4,c#5] parquet
!         +- Relation default.t1[a#3,b#4,c#5] parquet 
```

```
select a2 from (select a as a2, x as x2 from t1 join t2 on t1.a = t2.x) t where a2 = 1

=== Applying Rule org.apache.spark.sql.catalyst.optimizer.PushDownPredicates ===
 Project [a2#8]                                            Project [a2#8]
!+- Filter (a2#8 = 1)                                      +- Project [a#3 AS a2#8, x#10 AS x2#9]
!   +- Project [a#3 AS a2#8, x#10 AS x2#9]                    +- Join Inner, (a#3 = x#10)
!      +- Join Inner, (a#3 = x#10)                               :- Filter (a#3 = 1)
!         :- Relation default.t1[a#3,b#4,c#5] parquet            :  +- Relation default.t1[a#3,b#4,c#5] parquet
!         +- Relation default.t2[x#10,y#11,z#12] parquet         +- Relation default.t2[x#10,y#11,z#12] parquet
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -913,10 +913,17 @@ object CollapseRepartition extends Rule[LogicalPlan] {
       case (false, true) => if (r.numPartitions >= child.numPartitions) child else r
       case _ => r.copy(child = child.child)
     }
+    case r @ Repartition(_, _, p @ Project(_, child: RepartitionOperation)) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:916]: Is `Project` the only node to deal with? how about `Filter`/`Union`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:916]: This reminds me of `EliminateSorts`

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:916]: Not all `RepartitionOperation` can be removed. Sometimes repartition before joining and filtering to increase parallelism, especially before `BroadcastNestedLoopJoin`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:916]: `Project` can also contain expensive computation (e.g. UDF), why it's safe to remove repartition before `Project`?

**Comment by wangyum** [comment]
Close it because it is not a very common case

**Pull request closed without merging**